### PR TITLE
SY-4067: Strongly type logs

### PR DESCRIPTION
**Pull request by pjdotson**
# Issue Pull Request

## Linear Issue

[SY-4067](https://linear.app/synnax/issue/SY-4067/strongly-type-logs)

## Description

Strongly type logs

## Basic Readiness

- [x] I have performed a self-review of my code.
- [x] I have added relevant, automated tests to cover the changes.
- [x] I have updated documentation to reflect the changes.

<h3>Greptile Summary</h3>

This PR strongly types the log component end-to-end: the Go `Log` struct replaces its opaque `msgpack.EncodedJSON` blob with named fields (`Channels`, `RemoteCreated`, `TimestampPrecision`, `ShowChannelNames`, `ShowReceiptTimestamp`), a generated ORC codec handles binary storage, and matching TypeScript schemas (`channelEntryZ`, `logZ`) are generated from the Oracle definition. A two-step gorp boot migration lifts existing stored blobs through the legacy v0/v1 chain, and a new console state v2 migrates persisted `color: string` entries to `color.Color` tuples, fixing the previous color-truthiness render bug.

- **Go side**: `v55.Log` (blob) → typed `Log` via `MigrateLog`; the legacy chain (`legacy.go`, `legacy/v0`, `legacy/v1`) decodes historical JSON blobs and normalises out-of-set enum values; `parseColor` safely falls back to the zero color on malformed input; `imex` registration is removed entirely.
- **TypeScript/Console side**: `logStateZ` is now derived from `log.logZ.pick()`, unifying the aether state schema with the client-library type; `v2.stateZ` stores `color.Color` tuples and `v2.stateMigration` converts v1 hex strings via `color.colorZ.safeParse`; `stateFromLog` provides a clean projection from the server `Log` to the console `State`.
- **Tests**: New `codec_gen_test.go`, `migrate_test.go`, and `migrations.spec.ts` suites cover round-trip encoding, the full migration chain, colour edge cases, and the `stateFromLog` helper.

<h3>Confidence Score: 4/5</h3>

Safe to merge with awareness that log import/export is intentionally removed and Go zero-value enum normalization is deferred to a follow-up PR.

This is a large cross-stack refactor. The previously-flagged color-truthiness render bug and the logStateZ export compile error are both fixed. The console v1→v2 migration correctly converts empty-string colors to color.ZERO. The Go boot migration handles corrupt rows gracefully without blocking startup. The remaining deferred items (zero-value enum strings from Writer.Create, ShowChannelNames/ShowReceiptTimestamp Go zero values) are acknowledged by the team for a later PR. No new blocking defects were found in the changed code.

console/src/log/types/v2.ts (migration name reuse), core/pkg/service/log/migrate_test.go (unsafe type assertion in test helper), core/pkg/service/log/types.gen.go (Go zero-value booleans differ from Oracle-declared defaults).

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| pluto/src/log/aether/Log.ts | Refactors logState → logStateZ (now properly exported) and fixes the critical color truthiness guard; delegates channel config type to log.channelEntryZ, removing the local channelConfigZ definition. |
| client/ts/src/log/types.gen.ts | Replaces the opaque data record schema with strongly-typed ChannelEntry, TimestampConfig, and logZ fields; color is required (no default), timestamp uses .prefault() for graceful missing-field handling. |
| console/src/log/types/v2.ts | New v2 console state that uses color.Color instead of hex strings, with a migrateColor helper that safely falls back to color.ZERO for empty or malformed hex strings during v1→v2 migration. |
| console/src/log/types/v1.ts | Frozen v1 type: removes dependency on Log.channelConfigZ from pluto, inlines the schema locally, and exports STATE_MIGRATION_NAME / SLICE_MIGRATION_NAME constants used by the v2 migration. |
| core/pkg/service/log/migrate.go | MigrateLog chains AutoMigrateLog → legacy.MigrateData → logFromV1 for the gorp boot migration; silently drops the body on decode errors so a single corrupt row never blocks booting. |
| core/pkg/service/log/types.gen.go | Replaces the opaque msgpack.EncodedJSON Data blob with strongly-typed ChannelEntry, TimestampConfig, and top-level Log fields; Go zero values for bool fields differ from Oracle-declared defaults of true. |
| core/pkg/service/log/codec.gen.go | Generated ORC codec for the typed Log and ChannelEntry; uses nil-presence bit for the channels slice and CollectionLen (internally Uint32) for the element count, verified as consistent. |
| core/pkg/service/log/service.go | Removes ImEx registration and adds the two-step gorp boot migration (msgpack_to_orc → v55_lift_typed_log); log import/export is no longer registered with the imex service. |
| core/pkg/service/log/migrations/legacy/legacy.go | Version-dispatch entry point for the legacy blob migration; peeks the version field and walks the chain through v0→v1, treating nil blobs and missing version as v0. |
| core/pkg/service/log/migrations/legacy/v1/data.go | Frozen v1 data type with Normalize() that substitutes documented enum defaults for out-of-set values; Color stays as a raw string for the lift layer to parse. |
| console/src/log/types/index.ts | Wires in the v2 migrations alongside v1, re-exports stateFromLog, and updates anyStateZ union to include v2.stateZ first for precedence. |
| console/src/log/Log.tsx | useSyncComponent now explicitly passes typed fields to client.logs.create, and useLoadRemote uses stateFromLog(v) instead of casting v.data as State. |
| console/src/log/slice.ts | Renames setChannelConfig → setChannelEntry with the unified log.ChannelEntry payload shape; removes ChannelConfig/ZERO_CHANNEL_CONFIG re-exports now handled by v2. |
| core/pkg/service/log/migrate_test.go | Comprehensive MigrateLog tests including fixture-based v0/v1 round-trips; loadV55 helper uses a bare type assertion m["key"].(string) that would panic on malformed fixtures rather than failing gracefully. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["Gorp DB\n(v55 msgpack blob)"] -->|"msgpack_to_orc migration"| B["v55.Log\n{Key, Name, Data: EncodedJSON}"]
    B -->|"v55_lift_typed_log\n(MigrateLog)"| C{"legacy.MigrateData\n(version dispatch)"}
    C -->|"version=0.0.0 or empty"| D["v0.Data\n{Channels: []Key}"]
    C -->|"version=1.0.0"| E["v1.Data\n{Channels: []ChannelEntry (color: string)}"]
    D -->|"v1.Migrate"| E
    E -->|"logFromV1 + Normalize"| F["typed Log\n{Channels: []ChannelEntry\n (color: color.Color)}"]
    C -->|"error → drop body"| G["Log {Key, Name only}"]
    F --> H["Gorp DB\n(typed ORC)"]
    G --> H

    subgraph "Console State Migration"
    I["v1.State\n(color: string)"] -->|"v2.stateMigration\n(migrateColor)"| J["v2.State\n(color: color.Color)"]
    K["v0.State\n(channels: number[])"] -->|"v1.stateMigration"| I
    end

    H -->|"API GET /log/retrieve"| L["TS logZ.parse()"]
    L --> M["color.colorZ\n{r,g,b,a} → [r,g,b,a]"]
    M --> J
```

<details><summary><h3>Comments Outside Diff (5)</h3></summary>

1. `pluto/src/log/aether/Log.ts`, line 164-168 ([link](https://github.com/synnaxlabs/synnax/blob/3131eaf9025457343c1619c9e578547e894867f5/pluto/src/log/aether/Log.ts#L164-L168)) 

   <a href="#">P1</a> `cfg.color` is now a `color.Color` tuple (`[r, g, b, a]`), and all JavaScript arrays are truthy — including `color.ZERO = [0, 0, 0, 0]`. This means every channel with the default zero color also enters the block, setting `i.valueColors[key] = [0, 0, 0, 0]` (alpha=0). Because the downstream `const entryColor = valueColors[channelKey] ?? textColor` uses `??` (nullish coalescing, not `||`), the transparent zero color is preferred over `textColor`, making those channel entries render as invisible text in the log viewport. `Channels.tsx` already uses `!color.isZero(config.color)` as the canonical check; the same guard is needed here.

2. `pluto/src/log/aether/Log.ts`, line 164-168 ([link](https://github.com/synnaxlabs/synnax/blob/d464bfb5ffa0b80e6cc83bc8540deb1247809aec/pluto/src/log/aether/Log.ts#L164-L168)) 

   <a href="#">P1</a> `cfg.color` is now a `color.Color` tuple (`[r, g, b, a]`), and all JavaScript arrays are truthy — including `color.ZERO = [0, 0, 0, 0]`. Every channel with the default zero color enters this block, setting `i.valueColors[key] = [0, 0, 0, 0]` (alpha = 0). Because the downstream entry-color lookup uses `??` (nullish coalescing), the transparent zero color is preferred over `textColor`, making those channels render as invisible text. `Channels.tsx` already uses `!color.isZero(config.color)` as the canonical guard; the same check is needed here.

3. `pluto/src/log/aether/Log.ts`, line 54-80 ([link](https://github.com/synnaxlabs/synnax/blob/fbf325691f35f28b56eacdf624b75a86365157f9/pluto/src/log/aether/Log.ts#L54-L80)) 

   <a href="#">P0</a> **`logStateZ` is never exported — compile-time error**

   The PR renames all _consumer_ references from `logState` to `logStateZ` (class definition lines 135–137, `use.ts`, and the spec import), but the actual export on line 54 is still named `logState`. TypeScript will emit `Cannot find name 'logStateZ'` for every reference, preventing the build from succeeding. The spec's `logStateZ.parse(…)` (which asserts `color` parses to `[255, 0, 0, 1]`) also confirms this is intended to be a schema that uses `log.channelEntryZ` (with `color.colorZ`) rather than the local `channelEntryZ` (with `z.string()`), so the fix is more than a rename: `logStateZ` needs to be defined using `log.channelEntryZ` for its `channels` field so that the color transforms apply.

4. `core/pkg/service/log/service.go`, line 77-99 ([link](https://github.com/synnaxlabs/synnax/blob/7c0d57dd660946b724c8fb563675a08b47110d70/core/pkg/service/log/service.go#L77-L99)) 

   <a href="#">P1</a> **Log import/export silently dropped from API**

   `RegisterImportExporter(s)` was removed along with the `ImEx` config field, so the log service no longer registers itself as an importer/exporter. Any API request to import or export a log resource will now be dispatched to the imex service, find no handler for the `log` resource type, and return an error (or be silently ignored). The `imex_test.go` was deleted rather than updated, confirming no replacement path exists. If log import/export should be preserved — even in the new strongly-typed form — the `Service.Import`/`Service.Export` methods need to be reimplemented against the typed `Log` struct and re-registered here.

5. `pluto/src/log/aether/Log.ts`, line 171-175 ([link](https://github.com/synnaxlabs/synnax/blob/7c0d57dd660946b724c8fb563675a08b47110d70/pluto/src/log/aether/Log.ts#L171-L175)) 

   <a href="#">P1</a> `cfg.color` is now a `color.Color` tuple (`[r, g, b, a]`), and all JavaScript arrays are truthy — including `color.ZERO = [0, 0, 0, 0]`. Every channel configured with the default zero color enters this block, setting `i.valueColors[key] = [0, 0, 0, 0]` (alpha = 0). Because the downstream lookup uses `valueColors[channelKey] ?? textColor` (nullish coalescing), the transparent zero color is preferred over `textColor`, making those channels render as invisible text. `Channels.tsx` already uses `!color.isZero(config.color)` as the canonical guard.
</details>

<sub>Reviews (35): Last reviewed commit: ["Go back on Oracle syntax"](https://github.com/synnaxlabs/synnax/commit/3255ea9fcec9bac207734c59e1aae0d02b6f09dd) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=32561346)</sub>

### CODE REVIEW

File: core/pkg/service/log/migrate.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	out, err := AutoMigrateLog(ctx, old)
+	if err != nil {
+		return Log{}, err
+	}
+	if len(old.Data) == 0 {
+		return out, nil
+	}
+	d, err := legacy.MigrateData(old.Data)
+	if err != nil {
+		return out, nil
```
[the remark this answers is not in the document]
Comment (pjdotson) [reply core/pkg/service/log/migrate.go:41]: this is completely acceptable, we don't want to block booting up the core because of malformed data.

**Comment by pjdotson** [comment]
@greptile /rerun

File: core/pkg/service/log/types.gen.go [hunk 2]
```diff
@@ -13,12 +13,40 @@ package log
...
+	// Precision is the number of decimal digits to display. -1 means "use the log-level
+	// precision"; 17 is the maximum significant digits for a float64.
+	Precision int32 `json:"precision" msgpack:"precision"`
+	// Alias is a human-readable alias displayed in place of the channel name.
+	Alias string `json:"alias" msgpack:"alias"`
+	// Timestamp is the per-channel timestamp display configuration.
+	Timestamp TimestampConfig `json:"timestamp" msgpack:"timestamp"`
+}
+
 // Log is a timestamped event and message logging component. Logs display chronological
```
[the remark this answers is not in the document]
Comment (pjdotson) [reply core/pkg/service/log/types.gen.go:50]: This is alright for the moment. enum validation is going to be done more thoroughly and less hap-hazardly in a later PR.

**Comment by pjdotson** [comment]
@greptile /rerun noting the comment

**Comment by pjdotson** [comment]
@greptile /rerun

File: pluto/src/log/aether/Log.spec.ts [hunk 3]
```diff
@@ -24,6 +25,19 @@ import { TestFactory } from "@/telem/aether/test/factory";
 import { mockRenderContext } from "@/testutil/render";
 import { SYNNAX_DARK, SYNNAX_LIGHT, type Theme, themeZ } from "@/theming/base/theme";
 
+const entry = (
```
Comment (emilbon99) [inline pluto/src/log/aether/Log.spec.ts]: Any reason this isn't done by using `.parse` on the generated `ChannelEntry` schema? seems like it would correctly supply these defaults.

**Review by emilbon99 — CHANGES_REQUESTED** [review]
Approved after comment is addressed
